[BUG] Exception during agent run

DevoxxGenie 1.8.5 - 1.8.8

1. Configured Ollama as Provider
2. Set a Cloud Model like Minmax M3
3. Let it run a task in Agent Mode
4. The run will end with an exception:
An unexpected error occurred: ExecutionException - Error occurred during prompt execution Caused by: NullPointerException - Cannot invoke "java.lang.Integer.intValue()" because the return value of "dev.langchain4j.model.output.TokenUsage.inputTokenCount()" is null Please check the IDE log for more details.

I first tried the 1.8.5 there it crashed practically immediately and the ai couldn't get any task done.  a few days later when i saw 1.8.8 is out i tried it again and the ai currently can at leas complete the task but practically every chat answer ends in an error.

--- a/src/main/java/com/devoxx/genie/model/request/ChatMessageContext.java
+++ b/src/main/java/com/devoxx/genie/model/request/ChatMessageContext.java
@@ -81,9 +81,13 @@ public class ChatMessageContext {
 
     public void setTokenUsageAndCost(TokenUsage tokenUsage) {
         this.tokenUsage = tokenUsage;
-        if (this.tokenUsage != null) {
-            this.cost = (tokenUsage.inputTokenCount() * languageModel.getInputCost() +
-                tokenUsage.outputTokenCount() * languageModel.getOutputCost()) / 1_000_000.0;
+        if (this.tokenUsage != null && languageModel != null) {
+            // Some providers (e.g. cloud models in agent mode) return a non-null TokenUsage
+            // whose input/output counts are null. Guard against unboxing null Integers (#1149).
+            int inputTokens = tokenUsage.inputTokenCount() != null ? tokenUsage.inputTokenCount() : 0;
+            int outputTokens = tokenUsage.outputTokenCount() != null ? tokenUsage.outputTokenCount() : 0;
+            this.cost = (inputTokens * languageModel.getInputCost() +
+                outputTokens * languageModel.getOutputCost()) / 1_000_000.0;
         }
     }
 }

--- a/src/test/java/com/devoxx/genie/model/request/ChatMessageContextTest.java
+++ b/src/test/java/com/devoxx/genie/model/request/ChatMessageContextTest.java
@@ -1,0 +1,64 @@
+package com.devoxx.genie.model.request;
+
+import com.devoxx.genie.model.LanguageModel;
+import dev.langchain4j.model.output.TokenUsage;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+class ChatMessageContextTest {
+
+    private static LanguageModel model() {
+        return LanguageModel.builder()
+                .inputCost(3.0)
+                .outputCost(15.0)
+                .build();
+    }
+
+    @Test
+    void setTokenUsageAndCost_computesCostWhenCountsPresent() {
+        ChatMessageContext context = ChatMessageContext.builder()
+                .languageModel(model())
+                .build();
+
+        context.setTokenUsageAndCost(new TokenUsage(1_000_000, 1_000_000));
+
+        // (1_000_000 * 3 + 1_000_000 * 15) / 1_000_000 = 18.0
+        assertThat(context.getCost()).isEqualTo(18.0);
+    }
+
+    /**
+     * Reproduces issue #1149: some providers (e.g. when running a cloud model in agent
+     * mode) return a non-null TokenUsage whose inputTokenCount()/outputTokenCount() are
+     * null. Auto-unboxing those nulls during cost calculation threw a NullPointerException.
+     */
+    @Test
+    void setTokenUsageAndCost_doesNotThrowWhenTokenCountsAreNull() {
+        ChatMessageContext context = ChatMessageContext.builder()
+                .languageModel(model())
+                .build();
+
+        TokenUsage nullCounts = new TokenUsage(null, null, null);
+
+        assertThatCode(() -> context.setTokenUsageAndCost(nullCounts))
+                .doesNotThrowAnyException();
+
+        assertThat(context.getTokenUsage()).isSameAs(nullCounts);
+        assertThat(context.getCost()).isZero();
+    }
+
+    @Test
+    void setTokenUsageAndCost_handlesPartialNullCounts() {
+        ChatMessageContext context = ChatMessageContext.builder()
+                .languageModel(model())
+                .build();
+
+        // input present, output null
+        assertThatCode(() -> context.setTokenUsageAndCost(new TokenUsage(1_000_000, null)))
+                .doesNotThrowAnyException();
+
+        // (1_000_000 * 3 + 0 * 15) / 1_000_000 = 3.0
+        assertThat(context.getCost()).isEqualTo(3.0);
+    }
+}

